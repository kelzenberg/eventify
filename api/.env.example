# copy & rename this file to "local.env"
POSTGRES_URL=localhost
POSTGRES_USER=eventify
POSTGRES_PASSWORD=dbpassword

JWT_SIGNING_KEY=ThisIsAnUltraInsecureSigningKeyForJwtsPleaseDoNotUseThisInProductionAndGenerateANewOneThatHasAtLeast512bits