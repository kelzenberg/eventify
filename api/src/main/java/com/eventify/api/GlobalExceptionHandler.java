package com.eventify.api;

import com.eventify.api.exceptions.ErrorDetails;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.core.AuthenticationException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.server.ResponseStatusException;

import javax.servlet.http.HttpServletRequest;
import java.util.Date;

@ControllerAdvice
public class GlobalExceptionHandler {

    // specific exceptions

    @ExceptionHandler(ResponseStatusException.class)
    public ResponseEntity<ErrorDetails> handleResponseStatusException(ResponseStatusException exception, HttpServletRequest request) {
        ErrorDetails errorDetails = new ErrorDetails(
                new Date(),
                exception.getRawStatusCode(),
                exception.getStatus(),
                exception.getMessage()
        );
        return new ResponseEntity<>(errorDetails, errorDetails.getStatusMessage());
    }

    @ExceptionHandler(AccessDeniedException.class)
    public ResponseEntity<ErrorDetails> handleAccessDeniedException(AccessDeniedException exception, HttpServletRequest request) {
        ErrorDetails errorDetails = new ErrorDetails(
                new Date(),
                HttpStatus.FORBIDDEN.value(),
                HttpStatus.FORBIDDEN,
                exception.getMessage()
        );
        return new ResponseEntity<>(errorDetails, errorDetails.getStatusMessage());
    }

    @ExceptionHandler(AuthenticationException.class)
    public ResponseEntity<ErrorDetails> handleAuthenticationException(AuthenticationException exception, HttpServletRequest request) {
        ErrorDetails errorDetails = new ErrorDetails(
                new Date(),
                HttpStatus.UNAUTHORIZED.value(),
                HttpStatus.UNAUTHORIZED,
                exception.getMessage()
        );
        return new ResponseEntity<>(errorDetails, errorDetails.getStatusMessage());
    }

    // global exceptions

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorDetails> handleGlobalException(Exception exception, HttpServletRequest request) {
        ErrorDetails errorDetails = new ErrorDetails(
                new Date(),
                HttpStatus.INTERNAL_SERVER_ERROR.value(),
                HttpStatus.INTERNAL_SERVER_ERROR,
                exception.getMessage()
        );
        return new ResponseEntity<>(errorDetails, errorDetails.getStatusMessage());
    }
}
