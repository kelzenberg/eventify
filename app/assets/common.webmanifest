{
    "icons": [
        { "src": "/assets/Favicon_192.png", "type": "image/png", "sizes": "192x192" },
        { "src": "/assets/Favicon_512.png", "type": "image/png", "sizes": "512x512" }
    ]
}