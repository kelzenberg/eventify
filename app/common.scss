@use "sass:map";
@import "./bootstrap.custom.scss";

body {
    margin: 0px;
    font-family: 'Athiti', sans-serif;
}

.fancy-background {
    position: fixed;
    top: 0px;
    left: 0px;
    right: 0px;
    bottom: 0px;
    z-index: -5;
    overflow: hidden;
    background: linear-gradient(160deg, #F6CAA2, #F2B680);

    &::before {
        content: "";
        position: absolute;
        z-index: 0;
        top: 0px;
        left: 0px;
        width: 200vw;
        height: 30vh;
        transform: rotateZ(-20deg);
        transform-origin: 0% 100%;
        background: #f5ceaa;
        filter: blur(5px);
    }

    &::after {
        content: "";
        position: absolute;
        z-index: 0;
        bottom: 0px;
        right: 0px;
        width: 200vw;
        height: 30vh;
        transform: rotateZ(-20deg);
        transform-origin: 100% 0%;
        background: #F6CAA2;
        filter: blur(5px);
    }
}

.iconBox {
    display: inline-block;
    position: relative;
    border-radius: 13px;
    border: 1px solid map.get($theme-colors, "primary");
    width: 50px;
    height: 50px;
    background-repeat: no-repeat;
    background-position: 50% 50%;
    vertical-align: middle;

    img {
        width: 100%;
        height: 100%;
    }

    @each $color, $value in $theme-colors {
        &.iconBox-#{$color} {
            border-color: $value;
        }
    }

    &.iconBox-removable:hover::after {
        content: "";
        position: absolute;
        top: 0px;
        left: 0px;
        right: 0px;
        bottom: 0px;
        border-radius: 13px;
        background-color: rgba(255, 255, 255, 0.5);
        background-image: url("/assets/icons/remove.svg");
        background-repeat: no-repeat;
        background-position: 50% 50%;
        pointer-events: none;
    }
}

.hover-accent:hover {
    background-color: #ececec;
}