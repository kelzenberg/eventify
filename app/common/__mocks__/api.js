const api = jest.requireActual("../api");

const UnexpectedServerError = api.default.UnexpectedServerError;

const authenticate = jest.fn().mockResolvedValue({});
const getUserInfo = jest.fn().mockResolvedValue({});
const register = jest.fn().mockResolvedValue({});
const completeRegistration = jest.fn().mockResolvedValue({});
const getAllUserEvents = jest.fn().mockResolvedValue({});
const createEvent = jest.fn().mockResolvedValue({});
const getEvent = jest.fn().mockResolvedValue({});
const saveEvent = jest.fn().mockResolvedValue({});
const leaveEvent = jest.fn().mockResolvedValue({});
const inviteToEvent = jest.fn().mockResolvedValue({});
const bounceFromEvent = jest.fn().mockResolvedValue({});
const addExpenseSharingModule = jest.fn().mockResolvedValue({});
const addPaymentToExpenseSharing = jest.fn().mockResolvedValue({});
const deletePaymentFromExpenseSharing = jest.fn().mockResolvedValue({});

export {
    authenticate,
    getUserInfo,
    register,
    completeRegistration,
    getAllUserEvents,
    createEvent,
    getEvent,
    saveEvent,
    leaveEvent,
    inviteToEvent,
    bounceFromEvent,
    addExpenseSharingModule,
    addPaymentToExpenseSharing,
    deletePaymentFromExpenseSharing,
    UnexpectedServerError
};
