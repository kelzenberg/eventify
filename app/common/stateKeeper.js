import React from 'react';

export const UserContext = React.createContext(null);

export function isAuthenticated() {
    return localStorage.getItem("USER_TOKEN") != null;
}

export function getDisplayName() {
    return "Username";
}

export function setCredentials(token, userInfo) {
    localStorage.setItem("USER_TOKEN", token);
}

export function getAuthenticationToken() {
    return localStorage.getItem("USER_TOKEN");
}

// getUserInfo returns the stored user information or dummy data if none is set
export function getUserInfo() {
    try {
        let data = JSON.parse(localStorage.getItem("USER_INFO"));
        if (data == null) throw new Error();
        return data;
    } catch(e) {
        return {
            "id": "",
            "createdAt": "2021-06-02T00:06:45.528+00:00",
            "updatedAt": "2021-06-02T00:06:45.528+00:00",
            "email": "",
            "displayName": "",
            "authRole": ""
        };
    }
}

// maybeUpdateUserInfo takes a user info structure and checks if it matches the stored information.
// If the informations match it does nothing and returns false.
// If the new user information differs from the stored one it saves the new data and returns true.
export function maybeUpdateUserInfo(newUserInfo) {
    if(localStorage.getItem("USER_INFO") === null) {
        setUserInfo(newUserInfo);
        return true;
    }
    let oldUserInfo = getUserInfo();
    for(let key in newUserInfo) {
        if(newUserInfo[key] !== oldUserInfo[key]) {
            setUserInfo(newUserInfo);
            return true;
        }
    }
    return false;
}

function setUserInfo(userInfo) {
    localStorage.setItem("USER_INFO", JSON.stringify(userInfo));
}

export function clearLogin() {
    localStorage.removeItem('USER_TOKEN');
    localStorage.removeItem('USER_INFO');
}