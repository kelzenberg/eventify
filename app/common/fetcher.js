import * as stateKeeper from "./stateKeeper";

const apiHost = "https://localhost:8443/";

class FetchError extends Error {
    constructor(msg, code) {
        super(msg);
        this.code = code;
    }
}

function request(options) {
    if(typeof options.body == "object") {
        options.body = JSON.stringify(options.body);
    }
    var additionalHeaders = {};
    if(options.headers !== undefined) {
        additionalHeaders = options.headers
    }
    if(options.parameters != undefined) {
        options.path += "?" + new URLSearchParams(options.parameters).toString();
    }
    let authorization = null;
    if(stateKeeper.isAuthenticated()) {
        authorization = `Bearer ${stateKeeper.getAuthenticationToken()}`;
    }
    return new Promise(function(resolve, reject){
        fetch(apiHost + options.path, {
            credentials: 'omit',
            body: options.body,
            headers: {
                'accept': 'application/json',
                'content-type': 'application/json',
                'authorization': authorization,
                ...additionalHeaders
            },
            method: options.method,
            mode: 'cors',
            redirect: 'follow',
            referrer: 'no-referrer',
            keepalive: options.keepalive === true
        })
        // Check for errors
        .then(response => {
            if(response.ok || (options.allowedStatuses != undefined && options.allowedStatuses.includes(response.status)))Â {
                return Promise.resolve(response)
            } else if(response.status >= 400) {
                return Promise.reject(new FetchError(`request returned with status code ${response.status}`, response.status));
            }
        })
        // Decode content
        .then(response => {
            var contentType = response.headers.get("content-type");
            if(contentType == null) {
                if(options.expect != undefined && options.expect != "empty") {
                    console.warn(`fetcher: got response with content type '${contentType}', but expected '${options.expect}'.`);
                    return Promise.reject(new FetchError(`Invalid content-type '${contentType}', but expected '${options.expect}'`));
                }
                return Promise.resolve([null, response]);
            }
            if(contentType.includes("application/json")) {
                if(options.expect != undefined && options.expect != "json") {
                    console.warn(`fetcher: got response with content type '${contentType}', but expected '${options.expect}'.`);
                    return Promise.reject(new FetchError(`Invalid content-type '${contentType}', but expected '${options.expect}'`));
                }
                return response.json().then(data => {
                    return [data, response];
                })
            } else if(contentType.includes("text/html")) {
                if(options.expect != undefined && options.expect != "text") {
                    console.warn(`fetcher: got response with content type '${contentType}', but expected '${options.expect}'.`);
                    return Promise.reject(new FetchError(`Invalid content-type '${contentType}', but expected '${options.expect}'`));
                }
                return response.text().then(data => {
                    return [data, response];
                })
            }
        })
        // Parse data
        .then(([data, response]) => {
            resolve([data, response]);
        })
        // Handle remaining errors
        .catch(error => {
            if(error instanceof FetchError)
                reject(error);
            else {
                console.log(error);
                reject(new FetchError("fetcher encountered an unexpected error: " + error.message, -100))
            }
        });
    });
}

const status = {
    continue: 100,
    switchingProtocols: 101,
    processing: 102,
    earlyHints: 103,
    ok: 200,
    created: 201,
    accepted: 202,
    nonAuthorativeInformation: 203,
    noContent: 204,
    resetContent: 205,
    partialContent: 206,
    alreadyReported: 208,
    imUsed: 226,
    multipleChoice: 300,
    movedPermanently: 301,
    found: 302,
    seeOther: 303,
    notModified: 304,
    useProxy: 305,
    temporaryRedirect: 307,
    permanentRedirect: 308,
    badRequest: 400,
    unauthorized: 401,
    paymentRequired: 402,
    forbidden: 403,
    notFound: 404,
    methodNotAllowed: 405,
    notAcceptable: 406,
    proxyAuthenticationRequired: 407,
    requestTimeout: 408,
    conflict: 409,
    gone: 410,
    lengthRequired: 411,
    preconditionFailed: 412,
    payloadTooLarge: 413,
    uriTooLong: 414,
    unsupportedMediaType: 415,
    requestedRangeNotSatisfiable: 416,
    expectationFailed: 417,
    imATeapot: 418,
    misdirectedRequest: 421,
    unprocessedEntity: 422,
    locked: 423,
    failedDependency: 424,
    tooEarly: 425,
    upgradeRequired: 426,
    preconditionRequired: 428,
    tooManyRequests: 429,
    requestHeaderFieldsTooLarge: 431,
    unavailableForLegalReasons: 451,
    internalServerError: 500,
    notImplemented: 501,
    badGateway: 502,
    serviceUnavailable: 503,
    gatewayTimeout: 504,
    httpVersionNotSupported: 505,
    variantAlsoNegotiates: 506,
    insufficientStorage: 507,
    loopDetected: 508,
    notExtended: 510,
    networkAuthenticationRequired: 511
};

export default {
    FetchError,
    request,
    status
}