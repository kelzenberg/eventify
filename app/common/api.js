import fetcher from "./fetcher";
import * as stateKeeper from "./stateKeeper";

export class UnexpectedServerError extends Error {
    constructor(msg, code) {
        super(msg);
    }
}

export function authenticate(email, password) {
    return new Promise((resolve, reject) => {
        fetcher.request({
            method: "POST",
            path: "authenticate",
            expect: "json",
            body: {"email": email, "password": password},
            allowedStatuses: [fetcher.status.unauthorized]
        })
        .then(([responseData, rawResponse]) => {
            if(rawResponse.status === fetcher.status.unauthorized) {
                reject();
            } else {
                stateKeeper.setCredentials(responseData.token);
                resolve();
            }
        })
        .catch(err => {
            reject(new UnexpectedServerError(err));
        })
    });
}

export function getUserInfo() {
    return fetcher.request({method: "GET", path: "me", expect: "json"}).then(([response]) => response);
}

export function register(email, password, displayName) {
    let body = {
        "email": email,
        "password": password,
        "displayName": displayName
    };
    return fetcher.request({method: "POST", path: "register", expect: "json", body: body})
    .then(([responseData, rawResponse]) => {
        stateKeeper.setCredentials(responseData.token);
        return Promise.resolve();
    })
}

export function logout() {
    console.warn("TODO: api logout");
}

// =========================================== EVENTS =========================================== //

export function getAllUserEvents() {
    return fetcher.request({method: "GET", path: "me/events", expect: "json"}).then(([response]) => response);
}

export function getEvent(eventID) {
    return fetcher.request({method: "GET", path: `events/${eventID}`, expect: "json"}).then(([response]) => response);

    return Promise.resolve({
        "id": "22f7c93e-cfdc-491f-baa2-732fa298d8f6",
        "createdAt": "2021-06-02T15:42:16.478+00:00",
        "updatedAt": "2021-06-02T15:42:16.478+00:00",
        "title": "TestEvent 1",
        "description": "This is a test description",
        "startedAt": "2021-12-24T13:33:37.111+00:00",
        "endedAt": null,
        "expenseSharingModules": [
            {
                "id": "0f994c11-58ba-4718-ae18-5fe099177329",
                "createdAt": "2021-06-02T15:42:40.985+00:00",
                "updatedAt": "2021-06-02T15:42:40.985+00:00",
                "title": "Expense Sharing Module 1",
                "description": "This is a test description",
                "payments": [
                    {
                        "id": "e2ba702a-3713-4804-9b67-de5f9f99b3a8",
                        "createdAt": "2021-06-02T15:43:05.301+00:00",
                        "updatedAt": "2021-06-02T15:43:05.301+00:00",
                        "title": "Payment 1",
                        "amount": 13.37,
                        "payer": {
                            "id": "1798a949-1f72-45ff-b83c-d8e796add4cc",
                            "createdAt": "2021-05-29T17:01:31.716+00:00",
                            "updatedAt": "2021-05-29T17:01:31.716+00:00",
                            "email": "admin@test.de",
                            "displayName": "Admin",
                            "authRole": "ADMIN"
                        },
                        "shareType": "FIXED"
                    }
                ]
            }
        ]
    });
}

export function saveEvent(event) {
    return fetcher.request({method: "PUT", path: `events/${event.id}`, expect: "json"}).then(([response]) => response);
}