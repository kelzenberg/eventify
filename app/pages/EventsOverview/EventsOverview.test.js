import React from 'react';
import { default as snapshotRenderer } from 'react-test-renderer';
import {fireEvent, render, act} from '@testing-library/react';
import "@testing-library/jest-dom/extend-expect";
import { Router, MemoryRouter } from "react-router-dom";
import { createMemoryHistory } from "history";
import EventsOverviewPage from "./EventsOverview";
import BootstrapModalMock from "../../__mocks__/bootstrapModalMock";
import * as api from "../../common/api";

const Event = EventsOverviewPage.__RewireAPI__.__get__("Event");
const NewEvent = EventsOverviewPage.__RewireAPI__.__get__("NewEvent");

EventsOverviewPage.__Rewire__("Modal", BootstrapModalMock);
jest.mock("../../common/api");

let event = {
    id: "123456789",
    title: "Event Title",
    startedAt: new Date("2021-06-27T23:38:43.000Z"),
    endedAt: new Date("2021-06-27T23:38:43.000Z"),
    amountOfUsers: 5,
    description: "Event Description"
};

afterEach(() => {
    jest.clearAllMocks();
})

describe('Event Card', () => {
    test('Snapshot', () => {
        let component = snapshotRenderer.create(
            <MemoryRouter>
                <Event event={event}/>
            </MemoryRouter>
        );

        expect(component.toJSON()).toMatchSnapshot();
    });

    test('Click', () => {
        const history = createMemoryHistory();
        const {getByText} = render(
            <Router history={history}>
                <Event event={event}/>
            </Router>
        );

        fireEvent.click(getByText(event.title));
        expect(history.location.pathname).toBe(`/you/event/${event.id}`);
    });
})

describe('NewEvent Card', () => {
    let refreshEvents = jest.fn();

    afterEach(() => {
        refreshEvents.mockReset();
    });

    test('Snapshot', () => {
        let component = snapshotRenderer.create(
            <NewEvent refreshEvents={refreshEvents}/>
        );

        expect(component.toJSON()).toMatchSnapshot();
    });

    test('Dialog', () => {
        const {container, getByLabelText, getByText} = render(
            <NewEvent refreshEvents={refreshEvents}/>
        );
    
        // expect the dialog to be hidden
        expect(BootstrapModalMock).toHaveBeenCalledWith(expect.objectContaining({
            show: false,
            onHide: expect.any(Function)
        }), expect.anything());

        // click the card
        fireEvent.click(getByText("new event", {exact: false}));

        // expect the dialog to be visible now
        expect(BootstrapModalMock).toHaveBeenCalledWith(expect.objectContaining({
            show: true,
            onHide: expect.any(Function)
        }), expect.anything());

        // Test interaction
        fireEvent.change(getByLabelText("name",        {exact: false}), {target: {value: event.title      }});
        fireEvent.change(getByLabelText("description", {exact: false}), {target: {value: event.description}});
        fireEvent.change(getByLabelText("start",       {exact: false}), {target: {value: event.startedAt  }});

        // because the button starts a promise we need to also wait for this promise to finish
        // before we can check that all changes have occurred in the component.
        // This waiting has to be wrapped in a call to act() to tell react that we are expecting changes in the background.
        let apiPromise = Promise.resolve();
        api.createEvent.mockReturnValueOnce(apiPromise);

        fireEvent.click(getByText("Create")); // click the create button

        return act(() => {
            return apiPromise.then(() => {
                expect(api.createEvent).toHaveBeenCalledWith(event.title, event.description, event.startedAt);
                expect(refreshEvents).toHaveBeenCalled();
                return Promise.resolve();
            });
        });
    });
});
