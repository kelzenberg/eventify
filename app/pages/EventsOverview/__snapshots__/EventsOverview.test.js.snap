// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event Card Snapshot 1`] = `
<div
  className="card h-100"
>
  <div
    className="card-body p-4 eventCardBody"
    onClick={[Function]}
  >
    <h5
      className="card-title fw-bolder fs-4 mb-4"
    >
      Event Title
    </h5>
    <h6
      className="card-subtitle mb-2 text-muted"
    >
      <div
        className="d-flex"
      >
        <div
          className="text-left me-auto"
        >
          <span
            className="text-primary"
          >
            28.6.2021
          </span>
          <br />
          <span
            className="text-primary fw-light"
          >
            01:38 - 01:38
          </span>
        </div>
        <div
          className="text-end pt-2"
        >
          <span>
            5
          </span>
          <img
            alt="attendees"
            className="ms-1"
            src="/assets/icons/members.svg"
          />
        </div>
      </div>
    </h6>
    <p
      className="card-text"
    >
      Event Description
    </p>
  </div>
  <div
    className="card-footer"
  >
    <a
      href="/you/event/123456789"
      onClick={[Function]}
    >
      <button
        aria-label="Edit Event"
        className="btn btn-primary me-3"
      >
        Edit
      </button>
    </a>
    <button
      aria-label="Delete Event"
      className="btn btn-outline-secondary btn-icon"
    >
      <img
        src="/assets/icons/trash.svg"
      />
    </button>
  </div>
</div>
`;

exports[`NewEvent Card Snapshot 1`] = `
<div
  aria-label="Create New Event"
  className="card-outline h-100"
  onClick={[Function]}
>
  <div
    className="card-body p-4 h-100 d-flex justify-content-center align-items-center flex-direction-column"
  >
    <div
      className="text-center"
    >
      <span
        className="iconBox"
        style={
          Object {
            "backgroundImage": "url(/assets/icons/add.svg)",
          }
        }
      />
      <p
        className="mt-3 text-primary fw-bold"
      >
        CREATE NEW EVENT
      </p>
    </div>
  </div>
</div>
`;
